File: sorting-algorithms/binary-search/javascript/binarysearch.js

var array = generate();
bubbleSort(array);

function init() {
	document.getElementById("arrayText").innerHTML = array.join(', ');
}

function search() {

	let value = parseInt(document.getElementById("inputValue").value);
	let indexValue = binarySearch(array, value);
	document.getElementById("result").innerHTML = "Index of " + value + " is " + (indexValue != -1 ? indexValue : " not valid");
}

/* gera valore aleatórios */
function generate() {
	let randomArray = [];	
	for (let i = 0; i < MAX; i++) {
		randomArray.push( Math.floor(Math.random() * MAX));
	}
	return randomArray;
}

/* implementação do bubble sort algorithm */
function bubbleSort(array) {
	for (let i = 0; i < array.length - 1; i++) {
		for (let j = 0; j < array.length - i - 1; j++) {
			if (array[j] > array[j + 1]) {
				let aux = array[j];
				array[j] = array[j + 1];
				array[j + 1] = aux;
			}
		}
	}
}

function binarySearch(array, value) {
	return binarySearchAlgorithm(array, 0, array.length - 1, value);
}

/* implementação do binary search algorithm */
function binarySearchAlgorithm(array, min, max, value) {
	
	if (min > max)
		return -1;
	
	let mid = Math.floor((min + max) / 2);
	
	if (value == array[mid])
		return mid;
	
	if (value > array[mid])
		return binarySearchAlgorithm(array, mid + 1, max, value);
	else
		return binarySearchAlgorithm(array, min, mid - 1, value);
	
}

File: sorting-algorithms/binary-search/java/BinarySearch.java
package daviddev16;

import java.util.Arrays;

public class BinarySearch {

	public static void main(String[] args) {
		
		BubbleSort bubbleSort = new BubbleSort();
		BinarySearch binarySearch = new BinarySearch();
		
		int[] array = {0, 3, 1, 8, 5, 4, 10, 9, 8, 7, 6};
		bubbleSort.sort(array);
		System.out.println("Sorted array: " + Arrays.toString(array));
		

		int indexOfValue = binarySearch.search(array, 9);
		System.out.println("Index of 5 is " + indexOfValue);
		
	}
	
	public int search(int[] array, int value) {
		return binarySearchAlgorithm(array, 0, array.length - 1, value);
	}
	
	private static int binarySearchAlgorithm(int[] array, int min, int max, int value) {

		if (max < min)
			return -1;
		
		int mid = (max + min) / 2;
		
		if (value == array[mid])
			return mid;
		
		if (value > array[mid])
			return binarySearchAlgorithm(array, mid + 1, max, value);
		else
			return binarySearchAlgorithm(array, min, mid - 1, value);

	}

}
